File: back/src/main/java/com/back/global/security/oauth2/info/KakaoOAuth2UserInfo.java
package com.back.global.security.oauth2.info;

import java.util.Map;

public class KakaoOAuth2UserInfo implements OAuth2UserInfo {
    private final Map<String, Object> attributes;

    public KakaoOAuth2UserInfo(Map<String, Object> attributes) {
        this.attributes = attributes;
    }

    @Override
    public String getProviderId() {
        Object id = attributes.get("id");
        return id == null ? null : id.toString();
    }

    @Override
    public String getEmail() {
        Object email = getKakaoAccount().get("email");
        return email == null ? null : email.toString();
    }

    @Override
    public String getName() {
        Object profileObject = getKakaoAccount().get("profile");

        if (!(profileObject instanceof Map<?, ?> profile)) {
            return "카카오사용자";
        }

        Object nickname = profile.get("nickname");
        return nickname == null ? "카카오사용자" : nickname.toString();
    }

    private Map<?, ?> getKakaoAccount() {
        Object kakaoAccount = attributes.get("kakao_account");

        if (kakaoAccount instanceof Map<?, ?> map) {
            return map;
        }

        return Map.of();
    }
}

File: back/src/main/kotlin/com/back/global/security/oauth2/info/KakaoOAuth2UserInfo.kt
package com.back.global.security.oauth2.info

class KakaoOAuth2UserInfo(
    private val attributes: Map<String, Any>
) : OAuth2UserInfo {

    override val providerId: String?
        get() = attributes["id"]?.toString()

    override val email: String?
        get() = kakaoAccount["email"]?.toString()

    override val name: String
        get() {
            val profile = kakaoAccount["profile"] as? Map<*, *> ?: return "카카오사용자"
            return profile["nickname"]?.toString() ?: "카카오사용자"
        }

    override val isEmailVerified: Boolean
        get() = kakaoAccount["is_email_valid"] == true &&
            kakaoAccount["is_email_verified"] == true

    private val kakaoAccount: Map<*, *>
        get() = (attributes["kakao_account"] as? Map<*, *>) ?: emptyMap<Any, Any>()
}
